test_predictoor_batcher.py failures: "caller is not the owner", "transfer amount exceeds balance"

### Summary of issue

Follow README for backend devs up to the pytest section. Then:
```console
pytest pdr_backend/models/test/test_predictoor_batcher.py
```

Result summary:
```text
FAILED pdr_backend/models/test/test_predictoor_batcher.py::test_submit_truevals - web3.exceptions.ContractLogicError: execution reverted: VM Exception while processing transaction: revert Ownable: caller is not the owner
FAILED pdr_backend/models/test/test_predictoor_batcher.py::test_submit_truevals_contracts - web3.exceptions.ContractLogicError: execution reverted: VM Exception while processing transaction: revert Ownable: caller is not the owner
FAILED pdr_backend/models/test/test_predictoor_batcher.py::test_consume_multiple - web3.exceptions.ContractLogicError: execution reverted: VM Exception while processing transaction: revert ERC20: transfer amount exceeds balance
```

### Full traceback

```text
(venv) trentmc@tlm-macbook: ~/code/pdr-backend $ pytest pdr_backend/models/test/test_predictoor_batcher.py::test_submit_truevals  -s 
======================================================================================================== test session starts ========================================================================================================
platform darwin -- Python 3.11.4, pytest-7.4.0, pluggy-1.3.0
rootdir: /Users/trentmc/code/pdr-backend
configfile: pytest.ini
plugins: web3-6.9.0, env-1.0.1
collected 1 item                                                                                                                                                                                                                    

pdr_backend/models/test/test_predictoor_batcher.py Deployed NFT: 0x401512136592Df000b422F46f978De53F886f71D
Pair set to ETH/USDT in 0x343f3f8958f165b6d15ce6dccd6759d435f1df47db3fc53532a6b5c887944b02
base set to ETH in 0x62922423b35cbe0167be04f945fc2b176847024ffbdd6057af15033b12ba7a8d
quote set to USDT in 0x8f1e19bae74a7613df1d672e219908b6244414a1d016dc43546e418e39ab4696
source set to kraken in 0x9f90fd22736677a428dc378c2b7183b1f7f4803271f4fc5f9269c82dafee96b4
timeframe set to 5m in 0x2ba8af3c3f9654d84efd500794df1ebd3bb3b89d28682fd00de45a891b862996
Erc20Deployer set to 0xBE5449a6A97aD46c8558A3356267Ee5D2731ab5e in 0x054e98182e5a1d93d15065bc7695340b5b0002efe328468a2afb679af88b88d6
F

============================================================================================================= FAILURES ==============================================================================================================
_______________________________________________________________________________________________________ test_submit_truevals ________________________________________________________________________________________________________

predictoor_contract = <pdr_backend.models.predictoor_contract.PredictoorContract object at 0x105d47350>, web3_config = <pdr_backend.util.web3_config.Web3Config object at 0x106b2d750>
predictoor_batcher = <pdr_backend.models.predictoor_batcher.PredictoorBatcher object at 0x106c27fd0>

    def test_submit_truevals(
        predictoor_contract: PredictoorContract,
        web3_config: Web3Config,
        predictoor_batcher: PredictoorBatcher,
    ):
        current_epoch = predictoor_contract.get_current_epoch_ts()
    
        # fast forward time
        predictoor_contract.config.w3.provider.make_request(
            RPCEndpoint("evm_increaseTime"), [SECONDS_PER_EPOCH * 10]
        )
        predictoor_contract.config.w3.provider.make_request(RPCEndpoint("evm_mine"), [])
    
        end_epoch = current_epoch + SECONDS_PER_EPOCH * 10
    
        # get trueval for epochs
        epochs = list(range(current_epoch, end_epoch, SECONDS_PER_EPOCH))
        truevals = [True] * len(epochs)
        cancels = [False] * len(epochs)
    
        # add predictoor helper as ercdeployer
        erc721addr = predictoor_contract.erc721_addr()
        datanft = DataNft(web3_config, erc721addr)
        datanft.add_to_create_erc20_list(predictoor_batcher.contract_address)
    
        truevals_before = [
            predictoor_contract.contract_instance.functions.trueValues(i).call()
            for i in epochs
        ]
        for trueval in truevals_before:
            assert trueval is False
    
>       predictoor_batcher.submit_truevals(
            predictoor_contract.contract_address, epochs, truevals, cancels
        )

pdr_backend/models/test/test_predictoor_batcher.py:42: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pdr_backend/models/predictoor_batcher.py:61: in submit_truevals
    ).transact({"from": self.config.owner, "gasPrice": gasPrice})
venv/lib/python3.11/site-packages/web3/contract/contract.py:301: in transact
    return transact_with_contract_function(
venv/lib/python3.11/site-packages/web3/contract/utils.py:172: in transact_with_contract_function
    txn_hash = w3.eth.send_transaction(transact_transaction)
venv/lib/python3.11/site-packages/web3/eth/eth.py:363: in send_transaction
    return self._send_transaction(transaction)
venv/lib/python3.11/site-packages/web3/module.py:73: in caller
    result = w3.manager.request_blocking(
venv/lib/python3.11/site-packages/web3/manager.py:249: in request_blocking
    response = self._make_request(method, params)
venv/lib/python3.11/site-packages/web3/manager.py:188: in _make_request
    return request_func(method, params)
venv/lib/python3.11/site-packages/web3/middleware/signing.py:171: in middleware
    transaction = format_and_fill_tx(params[0])
cytoolz/functoolz.pyx:518: in cytoolz.functoolz.Compose.__call__
    ???
cytoolz/functoolz.pyx:263: in cytoolz.functoolz.curry.__call__
    ???
venv/lib/python3.11/site-packages/web3/_utils/transactions.py:132: in fill_transaction_defaults
    default_val = default_getter(w3, transaction)
venv/lib/python3.11/site-packages/web3/_utils/transactions.py:72: in <lambda>
    "gas": lambda w3, tx: w3.eth.estimate_gas(tx),
venv/lib/python3.11/site-packages/web3/eth/eth.py:293: in estimate_gas
    return self._estimate_gas(transaction, block_identifier)
venv/lib/python3.11/site-packages/web3/module.py:73: in caller
    result = w3.manager.request_blocking(
venv/lib/python3.11/site-packages/web3/manager.py:250: in request_blocking
    return self.formatted_response(
venv/lib/python3.11/site-packages/web3/manager.py:209: in formatted_response
    apply_error_formatters(error_formatters, response)
venv/lib/python3.11/site-packages/web3/manager.py:85: in apply_error_formatters
    formatted_resp = pipe(response, error_formatters)
cytoolz/functoolz.pyx:680: in cytoolz.functoolz.pipe
    ???
cytoolz/functoolz.pyx:655: in cytoolz.functoolz.c_pipe
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

response = {'error': {'code': -32000, 'data': {'hash': None, 'message': 'revert', 'programCounter': 290, 'reason': 'Ownable: call...cessing transaction: revert Ownable: caller is not the owner', 'name': 'RuntimeError', ...}, 'id': 3, 'jsonrpc': '2.0'}

    def raise_contract_logic_error_on_revert(response: RPCResponse) -> RPCResponse:
        """
        Reverts contain a `data` attribute with the following layout:
            "Reverted "
            Function selector for Error(string): 08c379a (4 bytes)
            Data offset: 32 (32 bytes)
            String length (32 bytes)
            Reason string (padded, use string length from above to get meaningful part)
    
        See also https://solidity.readthedocs.io/en/v0.6.3/control-structures.html#revert
        """
        if not isinstance(response["error"], dict):
            raise ValueError("Error expected to be a dict")
    
        data = response["error"].get("data", "")
        message = response["error"].get("message", "")
    
        message_present = message is not None and message != ""
    
        if data is None:
            if message_present:
                raise ContractLogicError(message)
            elif not message_present:
                raise ContractLogicError("execution reverted")
            else:
                raise Exception("Unreachable")
    
        # Ganache case:
        if isinstance(data, dict) and message_present:
>           raise ContractLogicError(f"execution reverted: {message}", data=data)
E           web3.exceptions.ContractLogicError: execution reverted: VM Exception while processing transaction: revert Ownable: caller is not the owner

venv/lib/python3.11/site-packages/web3/_utils/contract_error_handling.py:82: ContractLogicError
====================================================================================================== short test summary info ======================================================================================================
FAILED pdr_backend/models/test/test_predictoor_batcher.py::test_submit_truevals - web3.exceptions.ContractLogicError: execution reverted: VM Exception while processing transaction: revert Ownable: caller is not the owner
======================================================================================================== 1 failed in 10.11s =========================================================================================================
(venv) trentmc@tlm-macbook: ~/code/pdr-backend $ 
(venv) trentmc@tlm-macbook: ~/code/pdr-backend $ 
(venv) trentmc@tlm-macbook: ~/code/pdr-backend $ pytest pdr_backend/models/test/test_predictoor_batcher.py
======================================================================================================== test session starts ========================================================================================================
platform darwin -- Python 3.11.4, pytest-7.4.0, pluggy-1.3.0
rootdir: /Users/trentmc/code/pdr-backend
configfile: pytest.ini
plugins: web3-6.9.0, env-1.0.1
collected 3 items                                                                                                                                                                                                                   

pdr_backend/models/test/test_predictoor_batcher.py FFF                                                                                                                                                                        [100%]

============================================================================================================= FAILURES ==============================================================================================================
_______________________________________________________________________________________________________ test_submit_truevals ________________________________________________________________________________________________________

predictoor_contract = <pdr_backend.models.predictoor_contract.PredictoorContract object at 0x1021118d0>, web3_config = <pdr_backend.util.web3_config.Web3Config object at 0x106677250>
predictoor_batcher = <pdr_backend.models.predictoor_batcher.PredictoorBatcher object at 0x109085590>

    def test_submit_truevals(
        predictoor_contract: PredictoorContract,
        web3_config: Web3Config,
        predictoor_batcher: PredictoorBatcher,
    ):
        current_epoch = predictoor_contract.get_current_epoch_ts()
    
        # fast forward time
        predictoor_contract.config.w3.provider.make_request(
            RPCEndpoint("evm_increaseTime"), [SECONDS_PER_EPOCH * 10]
        )
        predictoor_contract.config.w3.provider.make_request(RPCEndpoint("evm_mine"), [])
    
        end_epoch = current_epoch + SECONDS_PER_EPOCH * 10
    
        # get trueval for epochs
        epochs = list(range(current_epoch, end_epoch, SECONDS_PER_EPOCH))
        truevals = [True] * len(epochs)
        cancels = [False] * len(epochs)
    
        # add predictoor helper as ercdeployer
        erc721addr = predictoor_contract.erc721_addr()
        datanft = DataNft(web3_config, erc721addr)
        datanft.add_to_create_erc20_list(predictoor_batcher.contract_address)
    
        truevals_before = [
            predictoor_contract.contract_instance.functions.trueValues(i).call()
            for i in epochs
        ]
        for trueval in truevals_before:
            assert trueval is False
    
>       predictoor_batcher.submit_truevals(
            predictoor_contract.contract_address, epochs, truevals, cancels
        )

pdr_backend/models/test/test_predictoor_batcher.py:42: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pdr_backend/models/predictoor_batcher.py:61: in submit_truevals
    ).transact({"from": self.config.owner, "gasPrice": gasPrice})
venv/lib/python3.11/site-packages/web3/contract/contract.py:301: in transact
    return transact_with_contract_function(
venv/lib/python3.11/site-packages/web3/contract/utils.py:172: in transact_with_contract_function
    txn_hash = w3.eth.send_transaction(transact_transaction)
venv/lib/python3.11/site-packages/web3/eth/eth.py:363: in send_transaction
    return self._send_transaction(transaction)
venv/lib/python3.11/site-packages/web3/module.py:73: in caller
    result = w3.manager.request_blocking(
venv/lib/python3.11/site-packages/web3/manager.py:249: in request_blocking
    response = self._make_request(method, params)
venv/lib/python3.11/site-packages/web3/manager.py:188: in _make_request
    return request_func(method, params)
venv/lib/python3.11/site-packages/web3/middleware/signing.py:171: in middleware
    transaction = format_and_fill_tx(params[0])
cytoolz/functoolz.pyx:518: in cytoolz.functoolz.Compose.__call__
    ???
cytoolz/functoolz.pyx:263: in cytoolz.functoolz.curry.__call__
    ???
venv/lib/python3.11/site-packages/web3/_utils/transactions.py:132: in fill_transaction_defaults
    default_val = default_getter(w3, transaction)
venv/lib/python3.11/site-packages/web3/_utils/transactions.py:72: in <lambda>
    "gas": lambda w3, tx: w3.eth.estimate_gas(tx),
venv/lib/python3.11/site-packages/web3/eth/eth.py:293: in estimate_gas
    return self._estimate_gas(transaction, block_identifier)
venv/lib/python3.11/site-packages/web3/module.py:73: in caller
    result = w3.manager.request_blocking(
venv/lib/python3.11/site-packages/web3/manager.py:250: in request_blocking
    return self.formatted_response(
venv/lib/python3.11/site-packages/web3/manager.py:209: in formatted_response
    apply_error_formatters(error_formatters, response)
venv/lib/python3.11/site-packages/web3/manager.py:85: in apply_error_formatters
    formatted_resp = pipe(response, error_formatters)
cytoolz/functoolz.pyx:680: in cytoolz.functoolz.pipe
    ???
cytoolz/functoolz.pyx:655: in cytoolz.functoolz.c_pipe
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

response = {'error': {'code': -32000, 'data': {'hash': None, 'message': 'revert', 'programCounter': 290, 'reason': 'Ownable: call...cessing transaction: revert Ownable: caller is not the owner', 'name': 'RuntimeError', ...}, 'id': 3, 'jsonrpc': '2.0'}

    def raise_contract_logic_error_on_revert(response: RPCResponse) -> RPCResponse:
        """
        Reverts contain a `data` attribute with the following layout:
            "Reverted "
            Function selector for Error(string): 08c379a (4 bytes)
            Data offset: 32 (32 bytes)
            String length (32 bytes)
            Reason string (padded, use string length from above to get meaningful part)
    
        See also https://solidity.readthedocs.io/en/v0.6.3/control-structures.html#revert
        """
        if not isinstance(response["error"], dict):
            raise ValueError("Error expected to be a dict")
    
        data = response["error"].get("data", "")
        message = response["error"].get("message", "")
    
        message_present = message is not None and message != ""
    
        if data is None:
            if message_present:
                raise ContractLogicError(message)
            elif not message_present:
                raise ContractLogicError("execution reverted")
            else:
                raise Exception("Unreachable")
    
        # Ganache case:
        if isinstance(data, dict) and message_present:
>           raise ContractLogicError(f"execution reverted: {message}", data=data)
E           web3.exceptions.ContractLogicError: execution reverted: VM Exception while processing transaction: revert Ownable: caller is not the owner

venv/lib/python3.11/site-packages/web3/_utils/contract_error_handling.py:82: ContractLogicError
------------------------------------------------------------------------------------------------------- Captured stdout setup -------------------------------------------------------------------------------------------------------
Deployed NFT: 0x71cb8aAad045436EC63e767dEb17a245701Ff4f3
Pair set to ETH/USDT in 0x9420a3d79fe36ac90032dccbe7a6e70a2ff38375580617f4022576242347cbc4
base set to ETH in 0xae6e2a8b803ce6e99b39559b4ab468da751df83182b488796c0adaa8a89fc3cc
quote set to USDT in 0x62124b24934597144f6c764c052f14ebd764f2254e4cbc8da4dd83347cb5dc78
source set to kraken in 0xc21c55415c1ff86ef1e6287e97ada70e9d2a31403c5e50ca6333d3a64f101a44
timeframe set to 5m in 0x4d60d94ce5eed6ad98cad2313804314a6cf9c96584ce7ccbd1b55baf12bb86c4
Erc20Deployer set to 0xBE5449a6A97aD46c8558A3356267Ee5D2731ab5e in 0x5f0afd2de4f8333dfc97f7887bd277033ed0cfe283a129baddc604bc2bd945db
__________________________________________________________________________________________________ test_submit_truevals_contracts ___________________________________________________________________________________________________

predictoor_contract = <pdr_backend.models.predictoor_contract.PredictoorContract object at 0x1021118d0>, predictoor_contract2 = <pdr_backend.models.predictoor_contract.PredictoorContract object at 0x1095f9a90>
web3_config = <pdr_backend.util.web3_config.Web3Config object at 0x106677250>, predictoor_batcher = <pdr_backend.models.predictoor_batcher.PredictoorBatcher object at 0x109085590>

    def test_submit_truevals_contracts(
        predictoor_contract: PredictoorContract,
        predictoor_contract2: PredictoorContract,
        web3_config: Web3Config,
        predictoor_batcher: PredictoorBatcher,
    ):
        current_epoch = predictoor_contract.get_current_epoch_ts()
    
        # fast forward time
        predictoor_contract.config.w3.provider.make_request(
            RPCEndpoint("evm_increaseTime"), [SECONDS_PER_EPOCH * 10]
        )
        predictoor_contract.config.w3.provider.make_request(RPCEndpoint("evm_mine"), [])
    
        end_epoch = current_epoch + SECONDS_PER_EPOCH * 10
    
        # get trueval for epochs
        epochs1 = list(range(current_epoch, end_epoch, SECONDS_PER_EPOCH))
        epochs2 = list(
            range(current_epoch + SECONDS_PER_EPOCH * 2, end_epoch, SECONDS_PER_EPOCH)
        )
        epochs = [epochs1, epochs2]
        truevals = [[True] * len(epochs1), [True] * len(epochs2)]
        cancels = [[False] * len(epochs1), [False] * len(epochs2)]
        addresses = [
            predictoor_contract.contract_address,
            predictoor_contract2.contract_address,
        ]
    
        # add predictoor helper as ercdeployer
        erc721addr = predictoor_contract.erc721_addr()
        datanft = DataNft(web3_config, erc721addr)
        datanft.add_to_create_erc20_list(predictoor_batcher.contract_address)
        erc721addr = predictoor_contract2.erc721_addr()
        datanft = DataNft(web3_config, erc721addr)
        datanft.add_to_create_erc20_list(predictoor_batcher.contract_address)
    
        truevals_before_1 = [
            predictoor_contract.contract_instance.functions.trueValues(i).call()
            for i in epochs1
        ]
    
        truevals_before_2 = [
            predictoor_contract2.contract_instance.functions.trueValues(i).call()
            for i in epochs1
        ]
    
        for trueval in truevals_before_1:
            assert trueval is False
    
        for trueval in truevals_before_2:
            assert trueval is False
    
>       predictoor_batcher.submit_truevals_contracts(addresses, epochs, truevals, cancels)

pdr_backend/models/test/test_predictoor_batcher.py:107: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pdr_backend/models/predictoor_batcher.py:45: in submit_truevals_contracts
    ).transact({"from": self.config.owner, "gasPrice": gasPrice})
venv/lib/python3.11/site-packages/web3/contract/contract.py:301: in transact
    return transact_with_contract_function(
venv/lib/python3.11/site-packages/web3/contract/utils.py:172: in transact_with_contract_function
    txn_hash = w3.eth.send_transaction(transact_transaction)
venv/lib/python3.11/site-packages/web3/eth/eth.py:363: in send_transaction
    return self._send_transaction(transaction)
venv/lib/python3.11/site-packages/web3/module.py:73: in caller
    result = w3.manager.request_blocking(
venv/lib/python3.11/site-packages/web3/manager.py:249: in request_blocking
    response = self._make_request(method, params)
venv/lib/python3.11/site-packages/web3/manager.py:188: in _make_request
    return request_func(method, params)
venv/lib/python3.11/site-packages/web3/middleware/signing.py:171: in middleware
    transaction = format_and_fill_tx(params[0])
cytoolz/functoolz.pyx:518: in cytoolz.functoolz.Compose.__call__
    ???
cytoolz/functoolz.pyx:263: in cytoolz.functoolz.curry.__call__
    ???
venv/lib/python3.11/site-packages/web3/_utils/transactions.py:132: in fill_transaction_defaults
    default_val = default_getter(w3, transaction)
venv/lib/python3.11/site-packages/web3/_utils/transactions.py:72: in <lambda>
    "gas": lambda w3, tx: w3.eth.estimate_gas(tx),
venv/lib/python3.11/site-packages/web3/eth/eth.py:293: in estimate_gas
    return self._estimate_gas(transaction, block_identifier)
venv/lib/python3.11/site-packages/web3/module.py:73: in caller
    result = w3.manager.request_blocking(
venv/lib/python3.11/site-packages/web3/manager.py:250: in request_blocking
    return self.formatted_response(
venv/lib/python3.11/site-packages/web3/manager.py:209: in formatted_response
    apply_error_formatters(error_formatters, response)
venv/lib/python3.11/site-packages/web3/manager.py:85: in apply_error_formatters
    formatted_resp = pipe(response, error_formatters)
cytoolz/functoolz.pyx:680: in cytoolz.functoolz.pipe
    ???
cytoolz/functoolz.pyx:655: in cytoolz.functoolz.c_pipe
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

response = {'error': {'code': -32000, 'data': {'hash': None, 'message': 'revert', 'programCounter': 290, 'reason': 'Ownable: call...cessing transaction: revert Ownable: caller is not the owner', 'name': 'RuntimeError', ...}, 'id': 7, 'jsonrpc': '2.0'}

    def raise_contract_logic_error_on_revert(response: RPCResponse) -> RPCResponse:
        """
        Reverts contain a `data` attribute with the following layout:
            "Reverted "
            Function selector for Error(string): 08c379a (4 bytes)
            Data offset: 32 (32 bytes)
            String length (32 bytes)
            Reason string (padded, use string length from above to get meaningful part)
    
        See also https://solidity.readthedocs.io/en/v0.6.3/control-structures.html#revert
        """
        if not isinstance(response["error"], dict):
            raise ValueError("Error expected to be a dict")
    
        data = response["error"].get("data", "")
        message = response["error"].get("message", "")
    
        message_present = message is not None and message != ""
    
        if data is None:
            if message_present:
                raise ContractLogicError(message)
            elif not message_present:
                raise ContractLogicError("execution reverted")
            else:
                raise Exception("Unreachable")
    
        # Ganache case:
        if isinstance(data, dict) and message_present:
>           raise ContractLogicError(f"execution reverted: {message}", data=data)
E           web3.exceptions.ContractLogicError: execution reverted: VM Exception while processing transaction: revert Ownable: caller is not the owner

venv/lib/python3.11/site-packages/web3/_utils/contract_error_handling.py:82: ContractLogicError
------------------------------------------------------------------------------------------------------- Captured stdout setup -------------------------------------------------------------------------------------------------------
Deployed NFT: 0xDD321681a9837F8a4a36214eE2F93F76d2099C42
Pair set to ETH/USDT in 0x3241d272e5bd6c89ec5c951e87bf270475f12f4c941bc5ee9ba6e127902b19a1
base set to ETH in 0x50a0c8f3201cf39d1822d0220913f2b4146f5eacf026f3367e66f833522e36bc
quote set to USDT in 0xf2a3a25ec7d91ca74751eeb7e369ab2aa5720d507c16e12574907aa14fbfbf8e
source set to kraken in 0xc9a6d145803d8aafe6893f4ceeacb022b71ace67755275cc351a31ce94620759
timeframe set to 5m in 0x8c45a99ba101d3cffa98cb2a248e601abf05bd9a5e2819b224a5c6ada65593c8
Erc20Deployer set to 0xBE5449a6A97aD46c8558A3356267Ee5D2731ab5e in 0xff15a897aa8d01bb07e21c2224b9880c2a435e25ba831b5b458a6348c1508fc1
_______________________________________________________________________________________________________ test_consume_multiple _______________________________________________________________________________________________________

predictoor_contract = <pdr_backend.models.predictoor_contract.PredictoorContract object at 0x1021118d0>, ocean_token = <pdr_backend.models.token.Token object at 0x109cd7610>
predictoor_batcher = <pdr_backend.models.predictoor_batcher.PredictoorBatcher object at 0x109085590>

    def test_consume_multiple(
        predictoor_contract: PredictoorContract,
        ocean_token: Token,
        predictoor_batcher: PredictoorBatcher,
    ):
        owner = ocean_token.config.owner
    
        price = predictoor_contract.get_price()
        print(price)
    
        times = 10
        cost = times * price
    
        ocean_token.approve(predictoor_batcher.contract_address, cost)
        balance_before = ocean_token.balanceOf(owner)
    
>       predictoor_batcher.consume_multiple(
            [predictoor_contract.contract_address], [times], ocean_token.contract_address
        )

pdr_backend/models/test/test_predictoor_batcher.py:145: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pdr_backend/models/predictoor_batcher.py:29: in consume_multiple
    ).transact({"from": self.config.owner, "gasPrice": gasPrice})
venv/lib/python3.11/site-packages/web3/contract/contract.py:301: in transact
    return transact_with_contract_function(
venv/lib/python3.11/site-packages/web3/contract/utils.py:172: in transact_with_contract_function
    txn_hash = w3.eth.send_transaction(transact_transaction)
venv/lib/python3.11/site-packages/web3/eth/eth.py:363: in send_transaction
    return self._send_transaction(transaction)
venv/lib/python3.11/site-packages/web3/module.py:73: in caller
    result = w3.manager.request_blocking(
venv/lib/python3.11/site-packages/web3/manager.py:249: in request_blocking
    response = self._make_request(method, params)
venv/lib/python3.11/site-packages/web3/manager.py:188: in _make_request
    return request_func(method, params)
venv/lib/python3.11/site-packages/web3/middleware/signing.py:171: in middleware
    transaction = format_and_fill_tx(params[0])
cytoolz/functoolz.pyx:518: in cytoolz.functoolz.Compose.__call__
    ???
cytoolz/functoolz.pyx:263: in cytoolz.functoolz.curry.__call__
    ???
venv/lib/python3.11/site-packages/web3/_utils/transactions.py:132: in fill_transaction_defaults
    default_val = default_getter(w3, transaction)
venv/lib/python3.11/site-packages/web3/_utils/transactions.py:72: in <lambda>
    "gas": lambda w3, tx: w3.eth.estimate_gas(tx),
venv/lib/python3.11/site-packages/web3/eth/eth.py:293: in estimate_gas
    return self._estimate_gas(transaction, block_identifier)
venv/lib/python3.11/site-packages/web3/module.py:73: in caller
    result = w3.manager.request_blocking(
venv/lib/python3.11/site-packages/web3/manager.py:250: in request_blocking
    return self.formatted_response(
venv/lib/python3.11/site-packages/web3/manager.py:209: in formatted_response
    apply_error_formatters(error_formatters, response)
venv/lib/python3.11/site-packages/web3/manager.py:85: in apply_error_formatters
    formatted_resp = pipe(response, error_formatters)
cytoolz/functoolz.pyx:680: in cytoolz.functoolz.pipe
    ???
cytoolz/functoolz.pyx:655: in cytoolz.functoolz.c_pipe
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

response = {'error': {'code': -32000, 'data': {'hash': None, 'message': 'revert', 'programCounter': 1149, 'reason': 'ERC20: trans... transaction: revert ERC20: transfer amount exceeds balance', 'name': 'RuntimeError', ...}, 'id': 11, 'jsonrpc': '2.0'}

    def raise_contract_logic_error_on_revert(response: RPCResponse) -> RPCResponse:
        """
        Reverts contain a `data` attribute with the following layout:
            "Reverted "
            Function selector for Error(string): 08c379a (4 bytes)
            Data offset: 32 (32 bytes)
            String length (32 bytes)
            Reason string (padded, use string length from above to get meaningful part)
    
        See also https://solidity.readthedocs.io/en/v0.6.3/control-structures.html#revert
        """
        if not isinstance(response["error"], dict):
            raise ValueError("Error expected to be a dict")
    
        data = response["error"].get("data", "")
        message = response["error"].get("message", "")
    
        message_present = message is not None and message != ""
    
        if data is None:
            if message_present:
                raise ContractLogicError(message)
            elif not message_present:
                raise ContractLogicError("execution reverted")
            else:
                raise Exception("Unreachable")
    
        # Ganache case:
        if isinstance(data, dict) and message_present:
>           raise ContractLogicError(f"execution reverted: {message}", data=data)
E           web3.exceptions.ContractLogicError: execution reverted: VM Exception while processing transaction: revert ERC20: transfer amount exceeds balance

venv/lib/python3.11/site-packages/web3/_utils/contract_error_handling.py:82: ContractLogicError
------------------------------------------------------------------------------------------------------- Captured stdout call --------------------------------------------------------------------------------------------------------
3603000000000000000
====================================================================================================== short test summary info ======================================================================================================
FAILED pdr_backend/models/test/test_predictoor_batcher.py::test_submit_truevals - web3.exceptions.ContractLogicError: execution reverted: VM Exception while processing transaction: revert Ownable: caller is not the owner
FAILED pdr_backend/models/test/test_predictoor_batcher.py::test_submit_truevals_contracts - web3.exceptions.ContractLogicError: execution reverted: VM Exception while processing transaction: revert Ownable: caller is not the owner
FAILED pdr_backend/models/test/test_predictoor_batcher.py::test_consume_multiple - web3.exceptions.ContractLogicError: execution reverted: VM Exception while processing transaction: revert ERC20: transfer amount exceeds balance
======================================================================================================== 3 failed in 21.69s =========================================================================================================
(venv) trentmc@tlm-macbook: ~/code/pdr-backend $ 
(venv) trentmc@tlm-macbook: ~/code/pdr-backend $ 
(venv) trentmc@tlm-macbook: ~/code/pdr-backend $ 
(venv) trentmc@tlm-macbook: ~/code/pdr-backend $ pytest pdr_backend/models/test/test_predictoor_batcher.py
```

--- a/READMEs/backend-dev.md
+++ b/READMEs/backend-dev.md
@@ -33,7 +33,7 @@ source venv/bin/activate
 export ADDRESS_FILE="${HOME}/.ocean/ocean-contracts/artifacts/address.json"
 export RPC_URL=http://127.0.0.1:8545
 export SUBGRAPH_URL="http://172.15.0.15:8000/subgraphs/name/oceanprotocol/ocean-subgraph"
-export PRIVATE_KEY="0xef4b441145c1d0f3b4bc6d61d29f5c6e502359481152f869247c7a4244d45209"
+export PRIVATE_KEY="0xc594c6e5def4bab63ac29eed19a134c130388f74f019bc74b8f4389df2837a58"
 ```
 
 Each subsection below gives different local usage options.

--- a/READMEs/predictoor.md
+++ b/READMEs/predictoor.md
@@ -40,7 +40,7 @@ source venv/bin/activate
 export ADDRESS_FILE="${HOME}/.ocean/ocean-contracts/artifacts/address.json"
 export RPC_URL=http://127.0.0.1:8545
 export SUBGRAPH_URL="http://172.15.0.15:8000/subgraphs/name/oceanprotocol/ocean-subgraph"
-export PRIVATE_KEY="0xef4b441145c1d0f3b4bc6d61d29f5c6e502359481152f869247c7a4244d45209"
+export PRIVATE_KEY="0xc594c6e5def4bab63ac29eed19a134c130388f74f019bc74b8f4389df2837a58"
 ```
 
 ### Local Usage: Random (Approach 1)

--- a/READMEs/trader.md
+++ b/READMEs/trader.md
@@ -39,7 +39,7 @@ source venv/bin/activate
 export ADDRESS_FILE="${HOME}/.ocean/ocean-contracts/artifacts/address.json"
 export RPC_URL=http://127.0.0.1:8545
 export SUBGRAPH_URL="http://172.15.0.15:8000/subgraphs/name/oceanprotocol/ocean-subgraph"
-export PRIVATE_KEY="0xef4b441145c1d0f3b4bc6d61d29f5c6e502359481152f869247c7a4244d45209"
+export PRIVATE_KEY="0xc594c6e5def4bab63ac29eed19a134c130388f74f019bc74b8f4389df2837a58"
 
 # run trader agent
 python pdr_backend/trader/main.py

--- a/pdr_backend/models/test/test_base_config.py
+++ b/pdr_backend/models/test/test_base_config.py
@@ -1,10 +1,11 @@
+import os
 from unittest.mock import patch, Mock
 
 from enforce_typing import enforce_types
 
 from pdr_backend.models.base_config import BaseConfig
 
-PRIV_KEY = "0xc594c6e5def4bab63ac29eed19a134c130388f74f019bc74b8f4389df2837a58"
+PRIV_KEY = os.getenv("PRIVATE_KEY")
 
 ADDR = "0xe8933f2950aec1080efad1ca160a6bb641ad245d"  # predictoor contract addr
 

--- a/pdr_backend/models/test/test_base_config.py
+++ b/pdr_backend/models/test/test_base_config.py
@@ -4,7 +4,7 @@ from enforce_typing import enforce_types
 
 from pdr_backend.models.base_config import BaseConfig
 
-PRIV_KEY = "0xef4b441145c1d0f3b4bc6d61d29f5c6e502359481152f869247c7a4244d45209"
+PRIV_KEY = "0xc594c6e5def4bab63ac29eed19a134c130388f74f019bc74b8f4389df2837a58"
 
 ADDR = "0xe8933f2950aec1080efad1ca160a6bb641ad245d"  # predictoor contract addr
 

--- a/pdr_backend/predictoor/approach1/test/test_predictoor_agent1.py
+++ b/pdr_backend/predictoor/approach1/test/test_predictoor_agent1.py
@@ -8,7 +8,7 @@ from pdr_backend.predictoor.approach1.predictoor_config1 import PredictoorConfig
 from pdr_backend.predictoor.approach1.predictoor_agent1 import PredictoorAgent1
 from pdr_backend.util.constants import S_PER_MIN, S_PER_DAY
 
-PRIV_KEY = "0xef4b441145c1d0f3b4bc6d61d29f5c6e502359481152f869247c7a4244d45209"
+PRIV_KEY = "0xc594c6e5def4bab63ac29eed19a134c130388f74f019bc74b8f4389df2837a58"
 
 ADDR = "0xe8933f2950aec1080efad1ca160a6bb641ad245d"
 

--- a/pdr_backend/predictoor/approach1/test/test_predictoor_agent1.py
+++ b/pdr_backend/predictoor/approach1/test/test_predictoor_agent1.py
@@ -1,3 +1,4 @@
+import os
 import random
 from typing import List
 from unittest.mock import Mock
@@ -8,7 +9,7 @@ from pdr_backend.predictoor.approach1.predictoor_config1 import PredictoorConfig
 from pdr_backend.predictoor.approach1.predictoor_agent1 import PredictoorAgent1
 from pdr_backend.util.constants import S_PER_MIN, S_PER_DAY
 
-PRIV_KEY = "0xc594c6e5def4bab63ac29eed19a134c130388f74f019bc74b8f4389df2837a58"
+PRIV_KEY = os.getenv("PRIVATE_KEY")
 
 ADDR = "0xe8933f2950aec1080efad1ca160a6bb641ad245d"
 

--- a/pdr_backend/predictoor/approach1/test/test_predictoor_config1.py
+++ b/pdr_backend/predictoor/approach1/test/test_predictoor_config1.py
@@ -7,6 +7,7 @@ from pdr_backend.predictoor.approach1.predictoor_config1 import PredictoorConfig
 ADDR = "0xe8933f2950aec1080efad1ca160a6bb641ad245d"  # predictoor contract addr
 PRIV_KEY = os.getenv("PRIVATE_KEY")
 
+
 @enforce_types
 def test_predictoor_config_basic(monkeypatch):
     _setenvs(monkeypatch)

--- a/pdr_backend/predictoor/approach1/test/test_predictoor_config1.py
+++ b/pdr_backend/predictoor/approach1/test/test_predictoor_config1.py
@@ -1,10 +1,11 @@
+import os
+
 from enforce_typing import enforce_types
 
 from pdr_backend.predictoor.approach1.predictoor_config1 import PredictoorConfig1
 
 ADDR = "0xe8933f2950aec1080efad1ca160a6bb641ad245d"  # predictoor contract addr
-PRIV_KEY = "0xc594c6e5def4bab63ac29eed19a134c130388f74f019bc74b8f4389df2837a58"
-
+PRIV_KEY = os.getenv("PRIVATE_KEY")
 
 @enforce_types
 def test_predictoor_config_basic(monkeypatch):

--- a/pdr_backend/predictoor/approach1/test/test_predictoor_config1.py
+++ b/pdr_backend/predictoor/approach1/test/test_predictoor_config1.py
@@ -3,7 +3,7 @@ from enforce_typing import enforce_types
 from pdr_backend.predictoor.approach1.predictoor_config1 import PredictoorConfig1
 
 ADDR = "0xe8933f2950aec1080efad1ca160a6bb641ad245d"  # predictoor contract addr
-PRIV_KEY = "0xef4b441145c1d0f3b4bc6d61d29f5c6e502359481152f869247c7a4244d45209"
+PRIV_KEY = "0xc594c6e5def4bab63ac29eed19a134c130388f74f019bc74b8f4389df2837a58"
 
 
 @enforce_types

--- a/pdr_backend/publisher/test/test_publish.py
+++ b/pdr_backend/publisher/test/test_publish.py
@@ -1,11 +1,14 @@
+import os
 from unittest.mock import Mock, call
+
+from eth_account import Account
 
 from pdr_backend.models.token import Token
 from pdr_backend.publisher.publish import fund_dev_accounts
 
 
 def test_fund_dev_accounts(monkeypatch):
-    pk = "0xef4b441145c1d0f3b4bc6d61d29f5c6e502359481152f869247c7a4244d45209"
+    pk = os.getenv("PRIVATE_KEY")
     monkeypatch.setenv("PREDICTOOR_PRIVATE_KEY", pk)
     monkeypatch.setenv("PREDICTOOR2_PRIVATE_KEY", pk)
 
@@ -19,9 +22,10 @@ def test_fund_dev_accounts(monkeypatch):
 
     fund_dev_accounts(accounts_to_fund, mock_account, mock_token)
 
+    address = Account.from_key(private_key=pk).address
     mock_token.transfer.assert_has_calls(
         [
-            call("0xBE5449a6A97aD46c8558A3356267Ee5D2731ab5e", 2e21, mock_account),
-            call("0xBE5449a6A97aD46c8558A3356267Ee5D2731ab5e", 3e21, mock_account),
+            call(address, 2e21, mock_account),
+            call(address, 3e21, mock_account),
         ]
     )

--- a/pdr_backend/publisher/test/test_publish.py
+++ b/pdr_backend/publisher/test/test_publish.py
@@ -5,7 +5,7 @@ from pdr_backend.publisher.publish import fund_dev_accounts
 
 
 def test_fund_dev_accounts(monkeypatch):
-    pk = "0xc594c6e5def4bab63ac29eed19a134c130388f74f019bc74b8f4389df2837a58"
+    pk = "0xef4b441145c1d0f3b4bc6d61d29f5c6e502359481152f869247c7a4244d45209"
     monkeypatch.setenv("PREDICTOOR_PRIVATE_KEY", pk)
     monkeypatch.setenv("PREDICTOOR2_PRIVATE_KEY", pk)
 

--- a/pdr_backend/publisher/test/test_publish.py
+++ b/pdr_backend/publisher/test/test_publish.py
@@ -5,7 +5,7 @@ from pdr_backend.publisher.publish import fund_dev_accounts
 
 
 def test_fund_dev_accounts(monkeypatch):
-    pk = "0xef4b441145c1d0f3b4bc6d61d29f5c6e502359481152f869247c7a4244d45209"
+    pk = "0xc594c6e5def4bab63ac29eed19a134c130388f74f019bc74b8f4389df2837a58"
     monkeypatch.setenv("PREDICTOOR_PRIVATE_KEY", pk)
     monkeypatch.setenv("PREDICTOOR2_PRIVATE_KEY", pk)
 

